open_spiel/games/phantom_go.cc

```diff
@@ -29,84 +29,178 @@ namespace {
+    {
+        newState->ApplyAction(VirtualActionToAction(kVirtualPass, boardSize));
+    }
+
+    for(long action : stones[player_id]) // Fill the board with stones of player we want to resample for
+    {
+        newState->ApplyAction(action);
+        newState->ApplyAction(VirtualActionToAction(kVirtualPass, boardSize));
+    }
+
+    if(!newState->history_.empty())
+    {
+        newState->UndoAction(-1, -1);
+    }
+
+
+    auto opp_player_id = (uint8_t)OppColor((GoColor)player_id);
+    for(long action : stones[opp_player_id])
+    {
+        newState->ApplyAction(action);
+        newState->ApplyAction(action);
+        newState->ApplyAction(VirtualActionToAction(kVirtualPass, boardSize));
+    }
+
+    for(int i = 0; i < stoneCount[opp_player_id] - stones[opp_player_id].size(); i++) {
+        std::vector<Action> actions = newState->LegalActions();
+        std::shuffle(actions.begin(), actions.end(), std::mt19937(std::random_device()()));
+        std::array<int, 2> currStoneCount = newState->board_.getStoneCount();
+        currStoneCount[opp_player_id]++;
+        std::vector<int> vec = stones[opp_player_id];
+
+        for (long action: actions) {
+            // pass can't be chosen, also an action that will be played by opposing player can't be chosen
+            if (action == VirtualActionToAction(kVirtualPass, boardSize) ||
+                std::find(vec.begin(), vec.end(), action) != vec.end())
+                continue;
+
+            newState->ApplyAction(action);
+            if (newState->board_.getStoneCount()[0] == currStoneCount[0] &&
+                newState->board_.getStoneCount()[1] == currStoneCount[1])
+            { //random move was applied correctly, no captures were made
+                newState->ApplyAction(VirtualActionToAction(kVirtualPass, boardSize));
+                //std::cout << "Added to observation " << ActionToString(c, action) << "\n";
+                break;
+            } else {
+                //std::cout << "random action" << ActionToString(c, action) << " was unacceptable\n";
+                newState->UndoAction(-1, -1);
+            }
+        }
+    }
+    if(!newState->history_.empty())
+    {
+        newState->UndoAction(-1, -1);
+    }
+
+    if (!(newState->board_.getStoneCount()[0] == stoneCount[0] &&
+        newState->board_.getStoneCount()[1] == stoneCount[1]))
+    {
+        ToString();
+        newState->ToString();
+        SpielFatalError("after resampling, the count of stones doesn't match\n");
+    }
+
+    if(CurrentPlayer() != newState->CurrentPlayer())
+    {
+        SpielFatalError("after resampling, wrong current player\n");
+    }
+
+    return newState;
 }
 
 
-// this method is in progress of making, the implementation is not correct
-std::unique_ptr<State> PhantomGoState::ResampleFromInfostate(
+std::unique_ptr<State> PhantomGoState::ResampleFromInfostateFull( //still to fix moves into eyes that keep messing up the histories
```

Comment:
This code seems complicated to review. Please use the feedback from the other method to improve your code and simplify it, and I can review it in the next iteration.

Reply:
To be honest, it is unfortunately pretty long, considering all the processes to be done to resample from Infostate. I will try to simplify as soon as the method is fully functinal. The hopefully last problem is with the random moves done by the opponent, that might create eyes on the points that the player will play in the future.